### ISSUE TITLE
run_name: deprecate? document intent, adjust DB schema, remove decision-making burden from users

### ISSUE THREAD

**Issue by austin3dickey** [conbench/conbench#607]
Now that run name is not used anywhere (except for a certain [picker](https://github.com/conbench/conbench/blob/main/conbench/templates/compare-list.html#L295) that we could change), we could consider deprecating it. All the projects using Conbench that I'm aware of populate run name by concatenating run reason and commit hash, both of which are easily accessible otherwise. It also can lead to [complications](https://github.com/voltrondata-labs/arrowbench/pull/127#discussion_r1080613598) in benchmark runners trying to provide a good UX with sensible defaults. It would reduce code complexity in conbench tools if we didn't have to worry about it.

What deprecation might look like:
- stop displaying it in the webapp
- if we need to uniquely, human-readably identify a run (in alerts or the webapp), we could auto-concatenate commit + hardware + reason (+ timestamp if necessary)
- have all conbench tools stop supplying it to the API when posting runs/results
- keep it around in the POST schema and database schema for backwards compatibility, with code comments

**Comment by jonkeane** [comment]
Can `run_name` currently be `""` or `null`?

**Comment by austin3dickey** [comment]
>Can run_name currently be "" or null?

Yes, it could be either. There are some older rows in Conbench databases we know about where this is the case.

(This currently may cause the aforementioned picker at the bottom of the `/compare/runs/A...B` page ([example](https://conbench.ursa.dev/compare/runs/8f36663cfd3f44e49405b9fe83811b7c...a0af231eecf746c8b60891054b1cf84c/)) to behave weirdly for those runs, but I'm not sure how widely-used that picker is, so it's never come up.)

This ticket is suggesting we go a step further than just allowing it to be empty, and actually remove most references to it, in order to reduce code complexity.

**Comment by jonkeane** [comment]
> This ticket is suggesting we go a step further than just allowing it to be empty, and actually remove most references to it, in order to reduce code complexity.

Yes, I understood the proposal. Could you say more about the code complexity we would be removing? Besides displaying it in the UI, and accepting it in `POST`s, and the picker, do we use it elsewhere (we shouldn't anymore after #507 yeah?)?

Ultimately, I think that going through a deprecation cycle for `run_name` is premature. While we have done good work to stop using some of the magic strings inside of it, having a user-specifiable place to put a name _is_ very helpful if anyone does any one off runs or tests. We don't currently support that setup very well, but if we ever did we would bring something like `run_name` back very very quickly. Having it around as a user-specifiable field that can be null, can be the same thing over and over, or whatever the conbench instance maintainer feels is helpful shouldn't be harmful elsewhere in our stack.  IMO it's not worth spending the energy now to deprecate and rip `run_name` out unless it's really hampering our ability to move forward other tickets (which it shouldn't anymore, and if it is we should figure out why that is and see if we can address the root-cause there).

**Comment by austin3dickey** [comment]
>Could you say more about the code complexity we would be removing?

Let me name some benefits I'm thinking about. You're right that these ultimately may not be worth it.

- In benchadapt, we could remove the `__post_init__` for [result](https://github.com/conbench/conbench/blob/680b6eaa2b4b6db1ac745c329b5f6963b4cf69e5/benchadapt/python/benchadapt/result.py#L137-L139) and [run](https://github.com/conbench/conbench/blob/680b6eaa2b4b6db1ac745c329b5f6963b4cf69e5/benchadapt/python/benchadapt/run.py#L91-L93). 
- In arrowbench, we could remove the [defaulting](https://github.com/voltrondata-labs/arrowbench/pull/127/files#diff-c118b736cd58f32580f07445199c10b7fec49b44f25f27cd46bbf1215cc9eff3R63) behavior too. We also wouldn't have to pass around the `CONBENCH_RUN_NAME` env var.
- User-facing cognitive load: it's just another argument to make a decision about when implementing benchmarks. We [recommend](https://github.com/conbench/conbench/blob/680b6eaa2b4b6db1ac745c329b5f6963b4cf69e5/conbench/entities/benchmark_result.py#L383) and/or [default](https://github.com/conbench/conbench/blob/680b6eaa2b4b6db1ac745c329b5f6963b4cf69e5/benchadapt/python/benchadapt/result.py#L17) in [a](https://github.com/conbench/conbench/blob/680b6eaa2b4b6db1ac745c329b5f6963b4cf69e5/benchadapt/python/benchadapt/run.py#L16) few [places](https://github.com/conbench/conbench/blob/680b6eaa2b4b6db1ac745c329b5f6963b4cf69e5/benchrun/python/benchrun/_benchmark_list.py#L32) to [use](https://github.com/conbench/conbench/blob/680b6eaa2b4b6db1ac745c329b5f6963b4cf69e5/conbench/runner.py#L65) `"{reason}: {commit}"` but that's not very useful (it doesn't uniquely identify runs (where's hardware?) and is redundant with other information)

>Besides displaying it in the UI, and accepting it in POSTs, and the picker, do we use it elsewhere?

Nope. If we fix the picker to truly uniquely ID runs, I'm not opposed to keeping `run_name` as is, and continuing to display it on the webapp benchmark page. I think what I actually dislike maintaining is the recommendation/default of `"{reason}: {commit}"`. A perfectly acceptable recommendation/default should be `None`.

**Comment by jonkeane** [comment]
> I think what I actually dislike maintaining is the recommendation/default of "{reason}: {commit}". A perfectly acceptable recommendation/default should be None.

💯 agreed that now that we don't rely on this (well mostly, the picker will probably need a bit of work there... but at least in the other cases where it mattered), removing that presumption | default from the runners would be great

**Comment by jgehrcke** [comment]
Thinking about the desired state (not current state). From a user's point of view. What would the main value be of being able to provide a custom run name? It's probably something related to discoverability/searchability. If we find clear value: maybe even require a run name (I am not a fan of having many _optional_ parameters).

If there is no clear value: let's remove the concept of run name.

**Comment by jonkeane** [comment]
> Thinking about the desired state (not current state). From a user's point of view. What would the main value be of being able to provide a custom run name? It's probably something related to discoverability/searchability. If we find clear value: maybe even require a run name (I am not a fan of having many optional parameters).

Yeah, the idea was that since runs are relatively numerous, being able to set them apart (especially for a manual run) makes it easy to find the thing you just triggered without needing to figure out how the triggering works or tracing that manually.

**Comment by jgehrcke** [comment]
The linked issues/PRs show that `run_name` creates work, while the value is still unclear to me.

Let's force ourselves to come up with a useful specification for `run_name` which really centers around the complete UX (how to use this for discovery). 

> makes it easy to find the thing you just triggered 

Do we have UI support for that as of today? If not: do we want to add this to the UI?

(I always love to simplify and remove primitives that are not fundamentally required)

**Comment by jonkeane** [comment]
IMO the biggest "feature" this has is that while one is testing a conbench orchestration or setting up a new one, you can add a memorable run name and then filter through the runs in the UI with the search bar to find it quickly. 

This will be much more important if we ever get back to supporting strongly one-off benchmark runs that are triggered manually. When we will almost certainly implement something like this again.

IMO we should decide if the kinda meh feature there + possibility of needing to implement the same thing all over again in the future outweighs the work to deprecate + remove this. Agreed that updating the API spec required work (though arguably that should have been done as part of the history project, since that was the project that most recently changed the meaning of run name). Now that that's done, do we expect it will cause us more work? And is that work greater than the work of ripping it out now? 

I'm not totally convinced that ripping it out now will save us much work. But I could be convinced (in which case let's do it swiftly and move on to more interesting things!)

**Comment by jgehrcke** [comment]
>  and then filter through the runs in the UI with the search bar to find it quickly.

Sounds like we have UI support for that today. But I do not know how to do that.

Does the uber-master search bar at the top right on the landing page support searching for run names? If yes: let's make this more obvious.

Update: I don't think it does.

I looked at a specific benchmark result and found this run name prefix:
![image](https://user-images.githubusercontent.com/265630/218084591-ca0ebc57-22b4-47e4-b2f9-254c082ae35b.png)

Then I went back to the landing page and typed this run name prefix into the top-right search bar:

![image](https://user-images.githubusercontent.com/265630/218084513-ce875fa7-bc30-4513-80a9-44f5310c93f9.png)

No results.

**Comment by jgehrcke** [comment]
The current spec for `run_name` reads (as served by `Conbench 2023.2.8-d7cf049ac` :mag: :rocket: ):

> Name for the Run (optional, does not need to be unique). Can be useful for implementing a custom naming convention. For organizing your benchmarks, and for enhanced search & discoverability. Ignored when Run was previously created.

That's what I had super quickly written without understanding exactly _how_ I would use this later on for "enhanced search & discoverability". This was after our brief exchange here: https://github.com/conbench/conbench/pull/628#discussion_r1088061054

I would appreciate if one of us could take a minute and propose an updated specification paragraph that makes it much more apparent how this is useful (written with a newbie incoming to this project in mind)!

If we like that we can then 
- actually update the spec with that
- build the API correspondingly
- adjust client tooling (docstrings, etc)
- potentially update the UI
- ...

**Comment by jonkeane** [comment]
> Update: I don't think it does.

Ack, I thought it did (and it might have in the past but probably lacked a test and we changed it so that it no longer did — I'm nearly certain I used it for exactly this when we were first setting things up). 

> I would appreciate if one of us could take a minute and propose an updated specification paragraph that makes it much more apparent how this is useful (written with a newbie incoming to this project in mind)!

I'll take this on and document it + make sure we at least have an issue for making it searchable again (or just do that myself)

**Comment by jgehrcke** [comment]
> propose an updated specification paragraph that makes it much more apparent how this is useful (written with a newbie incoming to this project in mind)!

New spec, proposal:
- db schema: string, maxlength (e.g. 100)
- not required, defaults to null/none.
- empty string not allowed (strictly)

I don't think we should have an index for this. I think the API should also not allow for filtering based on this. I think this should only be a brute-force client-side filter (in e.g. DataTables). 

New API spec/docs proposal:

> The `run_name` property is an optional legacy field that can be used for attaching a custom name to the Run you are about to create. This can help re-discovering this very Run via the Conbench UI later on. Note however that uniqueness is not enforced. This is for custom use cases, and our goal is that you don't miss out on a decent UX if you are fine not populating this field.

What do we think?

> make sure we at least have an issue for making it searchable again (or just do that myself)

Created https://github.com/conbench/conbench/issues/874.

> User-facing cognitive load: it's just another argument to make a decision about when implementing benchmarks.

I just want to corroborate this, this, this! That's why I find a "you don't need to worry about this if you don't know if you need this" part of docs really important. 

Will adjust the title to reflect the trajectory that this ticket is taking.

**Comment by jgehrcke** [comment]
> New API spec/docs proposal:

Anticipating that this is the direction we want to take I did some re-specification work here, two months ago: https://github.com/conbench/conbench/commit/3de257f7c85db5b6404416caa4da97c9f3099302#diff-9d75d1bf6ac23bfc056a410721cf4c9bccd8b302f8e3e83e4b4bf1698d0eaad7R409

TODO: reflect this nicely in code, and tweak those API docs.

**Comment by austin3dickey** [comment]
I like these ideas but think we should avoid terming this "legacy" (which to me implies deprecation). We wouldn't want users who want to use this feature to be worried that it will be removed someday.

**Comment by austin3dickey** [comment]
>(except for a certain [picker](https://github.com/conbench/conbench/blob/main/conbench/templates/compare-list.html#L295) that we could change)

This picker is gone in #1436.
